src/common/checksum.c

```diff
@@ -588,51 +588,168 @@ daos_csummer_alloc_iods_csums(struct daos_csummer *obj, daos_iod_t *iods,
+		daos_csummer_set_buffer(obj, buf, csum_info->cs_len);
+		daos_csummer_reset(obj);
+
+		chunk = csum_recx_chunkidx2range(recx, rec_len,
+						 rec_chunksize, i);
+
+		bytes_for_csum = chunk.dcr_nr * rec_len;
+		rc = daos_sgl_processor(sgl, idx, bytes_for_csum,
+					checksum_sgl_cb, obj);
+		if (rc != 0) {
+			D_ERROR("daos_sgl_processor error: %d\n", rc);
+			return rc;
+		}
+		daos_csummer_finish(obj);
+	}
+
+	return 0;
+}
+
+static bool
+range_overlaps(struct daos_csum_range *a, struct daos_csum_range *b)
+{
+	return (a->dcr_lo >= b->dcr_lo && a->dcr_lo <= b->dcr_hi) ||
+	       (b->dcr_lo >= a->dcr_lo && b->dcr_lo <= a->dcr_hi);
+}
+
+struct daos_csum_range
+get_maps_idx_nr_for_range(struct daos_csum_range *req_range, daos_iom_t *map)
+{
+	struct daos_csum_range	mapped_range;
+	struct daos_csum_range	result = {0};
+	daos_recx_t		mapped_recx;
+	int			i;
+
+	for (i = 0; i < map->iom_nr; i++) {
+		mapped_recx = map->iom_recxs[i];
+		/** turn mapped recx into a range */
+		dcr_set_idx_nr(&mapped_range, mapped_recx.rx_idx,
+			       mapped_recx.rx_nr);
+		if (range_overlaps(&mapped_range, req_range))
+			result.dcr_nr++;
+		else if (req_range->dcr_lo > mapped_range.dcr_hi)
+			result.dcr_lo++;
+	}
+
+	if (result.dcr_nr == 0) /** none found, reset idx */
+		result.dcr_lo = 0;
+	else
+		dcr_set_idx_nr(&result, result.dcr_lo, result.dcr_nr);
+
+	return result;
+}
+
+static int
+recx_hi(daos_recx_t *r)
+{
+	return r->rx_idx + r->rx_nr - 1;
+}
+
+static int
+calc_csum_recx_with_map(struct daos_csummer *obj, size_t csum_nr,
+			daos_recx_t *recx,
+			struct dcs_csum_info *csum_info,
+			daos_iom_t *map, size_t rec_len,
+			d_sg_list_t *sgl, uint32_t rec_chunksize,
+			struct daos_sgl_idx *idx)
+{
+	uint8_t			*buf;
+	struct daos_csum_range	 req_chunk;
+	struct daos_csum_range	 mapped_chunk;
+	int			 rc;
+	uint32_t		 i, j;
+	daos_size_t		 bytes_for_csum;
+	daos_size_t		 bytes_to_skip;
+	uint32_t		 prev_idx = recx->rx_idx;
+	struct daos_csum_range	 maps_in_chunk;
+
+	for (i = 0; i < csum_nr; i++) {
+		buf = ci_idx2csum(csum_info, i);
+		daos_csummer_set_buffer(obj, buf, csum_info->cs_len);
+		daos_csummer_reset(obj);
+		req_chunk = csum_recx_chunkidx2range(recx,
+						 rec_len, rec_chunksize,
+						 i);
+		maps_in_chunk = get_maps_idx_nr_for_range(&req_chunk, map);
+		for (j = maps_in_chunk.dcr_lo;
+		     j < maps_in_chunk.dcr_lo + maps_in_chunk.dcr_nr; j++) {
+			dcr_set_idxs(&mapped_chunk,
+				     max(req_chunk.dcr_lo,
+					 map->iom_recxs[j].rx_idx),
+				     min(req_chunk.dcr_hi,
+					 recx_hi(&map->iom_recxs[j])));
+
+			if (mapped_chunk.dcr_lo > prev_idx) {
+				bytes_to_skip = (mapped_chunk.dcr_lo - prev_idx)
+						* rec_len;
+				daos_sgl_get_bytes(sgl, idx,
+						   bytes_to_skip,
+						   NULL, NULL);
+			}
+			bytes_for_csum = mapped_chunk.dcr_nr * rec_len;
+			rc = daos_sgl_processor(sgl, idx, bytes_for_csum,
+						checksum_sgl_cb, obj);
+			if (rc != 0) {
+				D_ERROR("daos_sgl_processor error: %d\n", rc);
```

Comment:
In looking at the code, it appears nobody using a cf_finish callback....if one did, would you still want to call it on an error?

Reply:
Yes, good catch. Currently nothing is using cf_finish, but the sha checksum functions will. I'll make sure to fix this when I work on that.